node/src/container_chain_spawner.rs

```diff
@@ -124,208 +144,177 @@ pub enum CcSpawnMsg {
-        // This closure is used to emulate a try block, it enables using the `?` operator inside
-        let try_closure = move || async move {
-            // Preload genesis data from orchestrator chain storage.
-            // The preload must finish before calling create_configuration, so any async operations
-            // need to be awaited.
-
-            // TODO: the orchestrator chain node may not be fully synced yet,
-            // in that case we will be reading an old state.
-            let orchestrator_chain_info = orchestrator_client.chain_info();
-            log::info!(
-                "Reading container chain genesis data from orchestrator chain at block #{} {}",
-                orchestrator_chain_info.best_number,
-                orchestrator_chain_info.best_hash,
-            );
-            let orchestrator_runtime_api = orchestrator_client.runtime_api();
-
-            log::info!(
-                "Detected assignment for container chain {}",
-                container_chain_para_id
-            );
-
-            let genesis_data = orchestrator_runtime_api
-                .genesis_data(orchestrator_chain_info.best_hash, container_chain_para_id)
-                .map_err(|e| format!("Failed to call genesis_data runtime api: {}", e))?
-                .ok_or_else(|| {
-                    format!(
-                        "No genesis data registered for container chain id {}",
-                        container_chain_para_id
-                    )
-                })?;
-
-            let boot_nodes_raw = orchestrator_runtime_api
-                .boot_nodes(orchestrator_chain_info.best_hash, container_chain_para_id)
-                .map_err(|e| format!("Failed to call boot_nodes runtime api: {}", e))?;
-            if boot_nodes_raw.is_empty() {
-                log::warn!(
-                    "No boot nodes registered on-chain for container chain {}",
-                    container_chain_para_id
-                );
-            }
-            let boot_nodes =
-                parse_boot_nodes_ignore_invalid(boot_nodes_raw, container_chain_para_id);
-            if boot_nodes.is_empty() {
-                log::warn!(
-                    "No valid boot nodes for container chain {}",
-                    container_chain_para_id
-                );
-            }
-
-            container_chain_cli
-                .preload_chain_spec_from_genesis_data(
-                    container_chain_para_id.into(),
-                    genesis_data,
-                    chain_type.clone(),
-                    relay_chain.clone(),
-                    boot_nodes,
-                )
-                .map_err(|e| format!("failed to create container chain chain spec from on chain genesis data: {}", e))?;
+// Separate function to allow using `?` to return a result, and also to avoid using `self` in an
+// async function. Mutable state should be written by locking `state`.
+// TODO: `state` should be an async mutex
```

Comment:
I know but this refactor is already big enough, we were using the sync mutex already and it works good enough. To ensure no deadlocks happen I make sure to lock the mutex for the shortest amount of time possible, and never hold it across await points (the latter case doesn't even compile because MutexGuard is not Send).